File: net_sir_events.R
# R wrapper for the exact event-driven network SIR (net_sir_events.cpp).
#
# Drop-in alternative to run_stoch_adj() for the NETWORK model: same process,
# simulated exactly (no dt bias) in O(E log V) instead of O(n * max_degree)
# per dt step.
#
# Validity: time-homogeneous rates (scalar beta, constant gamma), waning = 0.
#
#   source("net_sir_events.R")   # compiles on first use

if (!requireNamespace("cpp11", quietly = TRUE))
  stop("net_sir_events.R needs the 'cpp11' package")

# Compile once per session (cheap: a single translation unit).
# cpp_source() injects the generated R wrapper into `env` and returns a
# DLLInfo, so take the function out of the environment rather than off the
# return value.
.net_sir_env <- new.env(parent = globalenv())
net_sir_compile <- function(force = FALSE) {
  if (!force && !is.null(.net_sir_env$net_sir_event_times))
    return(invisible(.net_sir_env$net_sir_event_times))
  cpp11::cpp_source("net_sir_events.cpp", env = .net_sir_env, quiet = TRUE)
  if (is.null(.net_sir_env$net_sir_event_times))
    stop("net_sir_event_times was not registered by cpp11::cpp_source()")
  invisible(.net_sir_env$net_sir_event_times)
}

# Contact matrix (or an existing adjacency list from contact_matrix_to_adj)
# -> 0-based CSR in the TRANSMIT direction.
#
# Direction matters: get_conact_matrix_pl returns an ASYMMETRIC (directed)
# matrix. In stoch_mod_adj.R the FOI on node i sums over neighbors[, i], i.e.
# row i lists who infects i -- a RECEIVE list. The event-driven simulator
# walks outward from an infectious node, so it needs the transpose: who node k
# can infect, = {i : k in recv(i)} = column k. Using the receive list to
# transmit silently simulates the reversed graph.
adj_to_csr <- function(contact_matrix = NULL, adj = NULL) {
  if (is.null(adj)) {
    n    <- nrow(contact_matrix)
    recv <- lapply(seq_len(n), function(i) which(contact_matrix[i, ] != 0))
  } else {
    n    <- ncol(adj$neighbors)
    recv <- lapply(seq_len(n), function(i) {
      k <- which(adj$mask[, i] == 1L)
      if (!length(k)) integer(0) else as.integer(adj$neighbors[k, i])
    })
  }
  # Transpose recv -> transmit: edge (k infects i) for each i, k in recv[[i]].
  to    <- rep.int(seq_len(n), lengths(recv))          # i (receiver)
  from  <- as.integer(unlist(recv, use.names = FALSE)) # k (transmitter)
  ord   <- order(from, method = "radix")
  deg   <- tabulate(from, nbins = n)
  list(n   = n,
       nbr = as.integer(to[ord]) - 1L,                 # 0-based receivers
       ptr = as.integer(c(0L, cumsum(deg))))
}

# Returns a data.table matching run_stoch_network()'s output: time, sim,
# C1 (cumulative unvaccinated cases), C2 (cumulative vaccinated cases).
#
# `beta` is on the SAME scale run_stoch_adj receives, i.e. run_stoch_network
# has already multiplied the user's beta by N / k_mean.
run_stoch_network_events <- function(beta, N, susceptibility = c(1, 1),
                                     t, vac, csr, gamma = 1,
                                     timepoints = NULL, I_ini = 2,
                                     transmissibility = NULL,
                                     n_sim = 100, seed = 1L, k_mean = 6) {
  net_sir_compile()
  if (is.null(timepoints)) timepoints <- seq(1, t, 1)
  n <- csr$n
  sus <- rep(susceptibility[1], n); sus[vac] <- susceptibility[2]
  tr  <- if (is.null(transmissibility)) rep(1, n) else transmissibility
  seeds <- seq_len(min(I_ini, n)) - 1L        # matches run_stoch_network's I_ini

  inf <- .net_sir_env$net_sir_event_times(
    n = as.integer(n), nbr = csr$nbr, ptr = csr$ptr,
    susceptibility = as.numeric(sus), transmissibility = as.numeric(tr),
    beta = as.numeric(N * beta / k_mean), gamma = as.numeric(gamma),
    t_max = as.numeric(max(timepoints)),
    seeds = as.integer(seeds), n_sim = as.integer(n_sim),
    seed = as.integer(seed))

  is_vac <- logical(n); is_vac[vac] <- TRUE
  data.table::rbindlist(lapply(timepoints, function(tp) {
    hit <- inf <= tp                          # [n_sim, n] cumulative incidence
    data.table::data.table(
      time = tp, sim = seq_len(n_sim),
      C1 = as.numeric(rowSums(hit[, !is_vac, drop = FALSE])),
      C2 = as.numeric(rowSums(hit[,  is_vac, drop = FALSE])))
  }))
}

File: net_sir_events.cpp
// Exact event-driven Markovian SIR on a static network.
//
// The dust tau-leaping model (stoch_mod_adj.R) evaluates every node at every
// dt step -- n * max_degree work per step -- however few events occur. For a
// sparse outbreak that is ~10^5 operations per realised infection. This
// simulates the SAME continuous-time process exactly, as first-passage
// percolation over the contact graph:
//
//   * a node confirmed infected at time t draws an infectious period
//     Exp(gamma), and for each not-yet-infected neighbour a transmission
//     delay Exp(rate); the transmission happens only if it lands inside the
//     infectious period;
//   * candidates go in a min-heap keyed on time, and the FIRST pop for a node
//     is its true infection time (later pops are stale and discarded).
//
// Rates match stoch_mod_adj.R exactly:
//     foi[i]  = beta / N_total * sum_{k in nbr(i)} I[k] * transmisibility[k]
//     hazard  = susceptibility[i] * foi[i]
//   so the per-edge rate k -> i is  beta / n * trans[k] * sus[i].
// `beta` here is the value passed to run_stoch_adj (run_stoch_network already
// scales the user's beta by N / k_mean).
//
// Cost is O(E log V) with no dt discretisation bias.
// Validity: time-homogeneous rates (scalar beta, constant gamma) and waning=0.
//
// Graph is CSR (nbr/ptr, 0-based), so work scales with EDGES, not
// n * max_degree.

#include <cpp11.hpp>
#include <vector>
#include <queue>
#include <random>
#include <cmath>
#include <limits>
#include <cstdint>
#include <algorithm>
#ifdef _OPENMP
#include <omp.h>
#endif

using namespace cpp11;

namespace {

struct Ev {
  double t;
  int node;
  bool operator>(const Ev& o) const { return t > o.t; }
};

// One realisation. inf[i] = infection time, or +Inf if never infected.
void one_sim(const int n,
             const std::vector<int>& nbr, const std::vector<int>& ptr,
             const std::vector<double>& sus, const std::vector<double>& trans,
             const double beta, const double gamma, const double t_max,
             const std::vector<int>& seeds,
             std::mt19937_64& rng,
             std::vector<double>& inf, std::vector<char>& done) {
  const double INF = std::numeric_limits<double>::infinity();
  std::fill(inf.begin(), inf.end(), INF);
  std::fill(done.begin(), done.end(), 0);
  std::exponential_distribution<double> exp1(1.0);
  std::priority_queue<Ev, std::vector<Ev>, std::greater<Ev>> pq;

  const double base = beta / static_cast<double>(n);   // matches foi's 1/N_total

  for (int s : seeds) if (s >= 0 && s < n) pq.push(Ev{0.0, s});

  while (!pq.empty()) {
    const Ev e = pq.top(); pq.pop();
    if (e.t > t_max) break;            // heap is time-ordered: nothing left in range
    const int i = e.node;
    if (done[i]) continue;             // stale candidate (infected earlier)
    done[i] = 1;
    inf[i]  = e.t;                     // first pop = true infection time

    // Infectious period, drawn once the infection is confirmed.
    const double rec_i = e.t + ((gamma > 0) ? exp1(rng) / gamma : INF);

    for (int k = ptr[i]; k < ptr[i + 1]; ++k) {
      const int j = nbr[k];
      if (done[j]) continue;
      const double rate = base * trans[i] * sus[j];
      if (rate <= 0) continue;
      const double t_ev = e.t + exp1(rng) / rate;      // first transmission i->j
      if (t_ev < rec_i && t_ev <= t_max) pq.push(Ev{t_ev, j});
    }
  }
}

}  // namespace

// n_sim x n matrix of infection times (Inf = never infected).
// nbr/ptr 0-based CSR; seeds 0-based node indices.
[[cpp11::register]]
doubles_matrix<> net_sir_event_times(int n,
                                     integers nbr, integers ptr,
                                     doubles susceptibility,
                                     doubles transmissibility,
                                     double beta, double gamma, double t_max,
                                     integers seeds, int n_sim, int seed,
                                     int n_threads) {
  const std::vector<int>    v_nbr(nbr.begin(), nbr.end());
  const std::vector<int>    v_ptr(ptr.begin(), ptr.end());
  const std::vector<double> v_sus(susceptibility.begin(), susceptibility.end());
  const std::vector<double> v_tr(transmissibility.begin(), transmissibility.end());
  const std::vector<int>    v_seed(seeds.begin(), seeds.end());

  // Simulate into a plain buffer so the loop can be threaded: each realisation
  // is independent, with its own RNG stream keyed on its index (so results do
  // NOT depend on the thread count or scheduling). R API calls are not
  // thread-safe, so the writable matrix is filled serially afterwards.
  std::vector<double> buf(static_cast<size_t>(n_sim) * n);

#ifdef _OPENMP
#pragma omp parallel for schedule(dynamic) num_threads(n_threads)
#endif
  for (int s = 0; s < n_sim; ++s) {
    std::vector<double> inf(n);
    std::vector<char>   done(n);
    std::mt19937_64 rng(static_cast<uint64_t>(seed) * 1000003ULL +
                        static_cast<uint64_t>(s) * 7919ULL + 1ULL);
    one_sim(n, v_nbr, v_ptr, v_sus, v_tr, beta, gamma, t_max, v_seed, rng,
            inf, done);
    std::copy(inf.begin(), inf.end(), buf.begin() + static_cast<size_t>(s) * n);
  }

  writable::doubles_matrix<> out(n_sim, n);
  for (int s = 0; s < n_sim; ++s)
    for (int i = 0; i < n; ++i) out(s, i) = buf[static_cast<size_t>(s) * n + i];
  return out;
}
